### PR TITLE
Add Wdf* APIs required by ebpf-for-windows

### LINKED ISSUE #31: Add cache-aligned allocation
* ExAllocatePoolUninitialized(NonPagedPoolNxCacheAligned, ...) -> _aligned_malloc
* ExFreePool(...) -> _aligned_free

### LINKED ISSUE #70: Missing Etw* APIs
The following are needed by ebpf-for-windows:
- [ ] EtwWriteTransfer (stub exists as of PR #73)
- [x] EtwRegister
- [x] EtwUnregister
- [ ] EtwSetInformation (stub exists as of PR #73)

### PR DESCRIPTION
**Pull request by dthaler**
* Add Wdf* APIs required by ebpf-for-windows (fixes #70)
* Fix bug in ke.cpp where wrong function was used to free memory allocated by C++ allocator
* Add cache-aligned allocations (fixes #31)
* Add stub for IofCompleteRequest (addresses part of #69)
* Add error check for freeing unallocated memory in leak_detector.cpp
* Move usersim_allocate functions from platform_user.cpp to ex.cpp where the other allocation functions are
* Make Ndis* and usersim_fwp_* functions callable from C code
* Fix bug in dll skeleton where DRIVER_OBJECT was freed during driver entry because it was on the stack.  It needs to survive until unload.

### CODE REVIEW

File: inc/usersim/ex.h [hunk 1]
```diff
@@ -198,6 +198,80 @@ USERSIM_API _Ret_maybenull_ void*
...
 
+/**
+ * @brief Allocate memory.
+ * @param[in] pool_type Pool type to use.
+ * @param[in] size Size of memory to allocate.
+ * @param[in] tag Pool tag to use.
+ * @param[in] initialize False to return "uninitialized" memory.
+ * @returns Pointer to memory block allocated, or null on failure.
+ */
+__drv_allocatesMem(Mem) _Must_inspect_result_ _Ret_writes_maybenull_(size) void* usersim_allocate_with_tag(
```
Comment (saxena-anurag) [inline inc/usersim/ex.h:209]: nit: what is the difference between `usersim_allocate_with_tag()` with PoolType `NonPagedPoolNxCacheAligned` and `usersim_allocate_cache_aligned()`?

Comment (dthaler) [reply inc/usersim/ex.h:209]: Haven't done cleanup, but usersim_allocate_cache_aligned doesn't do anything with the tag, so I think it should be removed.  usersim_allocate_cache_aligned_with_tag and usersim_allocate_with_tag() with PoolType NonPagedPoolNxCacheAligned are similar, but usersim_allocate_with_tag should really detect any cache-aligned pool type (it doesn't yet).  I suspect in the DLL form, nothing calls usersim_allocate_cache_aligned or usersim_allocate_cache_aligned_with_tag, but will check.

Comment (dthaler) [reply inc/usersim/ex.h]: I will clean this up in a separate PR as part of the work for https://github.com/microsoft/ebpf-for-windows/issues/2677.  In the meantime, I have added some TODO comments in the code.

File: src/platform_user.cpp [hunk 2]
```diff
@@ -204,9 +201,9 @@ usersim_platform_initiate()
     try {
         _usersim_platform_maximum_group_count = GetMaximumProcessorGroupCount();
         _usersim_platform_maximum_processor_count = GetMaximumProcessorCount(ALL_PROCESSOR_GROUPS);
-        auto fault_injection_stack_depth =
+        bool fault_injection_stack_depth =
```
Comment (Alan-Jowett) [inline src/platform_user.cpp:204]: This is size_t, not bool.

**Review by Alan-Jowett — APPROVED** [review]

**Review by mtfriesen — APPROVED** [review]

File: src/platform_user.cpp (hunk 2 above)
Comment (dthaler) [reply src/platform_user.cpp]: Fixed, thanks

**Pull request merged**